File: week-1/surrounded-regions/surrounded-regions.js
/**
 * @param {character[][]} board
 * @return {void} Do not return anything, modify board in-place instead.
 */

var solve = function(board) {
    if (board.length == 0) return board;
    let update = function (i, j, board, visited, char) {
        if (i<0 || j<0 || i>board.length-1 || j>board[0].length-1 || board[i][j] === 'X' || visited[i][j])
            return;
        board[i][j] = char;
        visited[i][j] = true;
        update(i+1, j, board, visited, char);
        update(i-1, j, board, visited, char);
        update(i, j+1, board, visited, char);
        update(i, j-1, board, visited, char);
    };
    
    for (let i=0; i<board.length; i++) {
        for (let j=0; j<board[i].length; j++) {
            if (i===0 || i===board.length-1 || j===0 || j===board[i].length-1) {
                let visited = new Array(board.length).fill(false).map(x => Array(board[i].length).fill(false));
                if (board[i][j] === 'O') update(i, j, board, visited, '1')
            }
        }
    }
          
    for (let i=1; i<board.length-1; i++) {
        for (let j=1; j<board[i].length-1; j++) {
            if (board[i][j] === 'O') board[i][j] = 'X';
        }
    }
        
    
    for (let i=0; i<board.length; i++) {
        for (let j=0; j<board[i].length; j++) {
            if (board[i][j] === '1') board[i][j] = 'O';
        }
    }
        
    return board;
};

File: week-1/surrounded-regions/surrounded-regions.py
class Solution:
    def solve(self, board: List[List[str]]) -> None:
        """
        Do not return anything, modify board in-place instead.
        """
        visited = [[False for j in i] for i in board]
        for i in range(0, len(board)):
            for j in range(0, len(board[i])):
                if board[i][j] == 'O' and (i == 0 or i == len(board)-1 or j == 0 or j == len(board[i])-1):
                    self.update(i, j, board, visited, '1')
        
        visited = [[False for j in i] for i in board]
        for i in range(1, len(board)-1):
            for j in range(1, len(board[i])-1):
                if board[i][j] == 'O':
                    self.update(i, j, board, visited, 'X')
        
        for i in range(0, len(board)):
            for j in range(0, len(board[i])):
                if board[i][j] == '1':
                    board[i][j] = 'O'
        
        
    def update(self, row, col, board, visited, character):
        if row<0 or col<0 or row>=len(board) or col>=len(board[0]) or visited[row][col] or board[row][col] == 'X':
            return
        
        visited[row][col] = True
        board[row][col] = character
        self.update(row+1, col, board, visited, character)
        self.update(row-1, col, board, visited, character)
        self.update(row, col+1, board, visited, character)
        self.update(row, col-1, board, visited, character)
